### PR TITLE
Split Expression Factory into separate translation units

### PR DESCRIPTION
**Pull request by klecki**
#### Why we need this PR?
- Refactoring to improve compilation time

#### What happened in this PR?
 - What solution was applied:

The template `ExprImplFactory*Op` factory function was
modified to only go over possible types. Its instantiation
is forced by invoking it only in and overload of `OpFactory`
which is implemented per arithm_meta
(so there is an overload for every ArithmeticOp x Backend pair).

The `OpFactory` overloads are defined in separate translation
units. It forces instantiating the implementations
in separate translation units and allows for
parallel compilation.

When compiling for one backed it reduces the
compilation time by half.

 - Affected modules and functionalities:

Implementation of expressions is changed.
All cc and cu files in expressions/expression_factory_instances/* are generated semi-automatically,
and most of the implementation is in dali/operators/expressions/expression_factory_instances/expression_impl_factory.h
The rest is just copyright notice.

 - Key points relevant for the review:

dali/operators/expressions/expression_factory_instances/expression_impl_factory.h

 - Validation and testing:

Local build time

 - Documentation (including examples):
N/A

**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by szalpal** [comment]
I'd suggest putting this:
```
ExprImplBase *OpFactory(arithm_meta<ArithmeticOp::add, CPUBackend>,
                        const ExprFunc &expr) {
  ExprImplBase *ret = nullptr;
  auto left_type = expr[0].GetTypeId();
  auto right_type = expr[1].GetTypeId();
  TYPE_SWITCH(left_type, type2id, Left_t, ARITHMETIC_ALLOWED_TYPES, (
    TYPE_SWITCH(right_type, type2id, Right_t, ARITHMETIC_ALLOWED_TYPES, (
          using Out_t = arithm_meta<ArithmeticOp::add, CPUBackend>::result_t<Left_t, Right_t>;
          if (expr[0].GetNodeType() == NodeType::Tensor && IsScalarLike(expr[1])) {
            ret = new ExprImplCpuTC<ArithmeticOp::add, Out_t, Left_t, Right_t>();
          } else if (IsScalarLike(expr[0]) && expr[1].GetNodeType() == NodeType::Tensor) {
            ret = new ExprImplCpuCT<ArithmeticOp::add, Out_t, Left_t, Right_t>();
          } else if (expr[0].GetNodeType() == NodeType::Tensor &&
                      expr[1].GetNodeType() == NodeType::Tensor) {
            ret = new ExprImplCpuTT<ArithmeticOp::add, Out_t, Left_t, Right_t>();
          } else {
            DALI_FAIL("Expression cannot have two scalar operands");
          }
    ), DALI_FAIL("No suitable type found"););  // NOLINT(whitespace/parens)
  ), DALI_FAIL("No suitable type found"););  // NOLINT(whitespace/parens)
  return ret;
}
```
In a macro. Reasons are numerous, e.g. it won't need 2kLOC commit, whenever something changes we don't need to make another 2kLOC commit, we already do this kind of stuff in macros...

**Comment by klecki** [comment]
> I'd suggest putting this:
> 
> ```
> ExprImplBase *OpFactory(arithm_meta<ArithmeticOp::add, CPUBackend>,
>                         const ExprFunc &expr) {
....
> ```
> 
> In a macro. Reasons are numerous, e.g. it won't need 2kLOC commit, whenever something changes we don't need to make another 2kLOC commit, we already do this kind of stuff in macros...

I would put it in a macro if the compiler would handle the TYPE_SWITCH in a sane way. I am yet to find how to do it.

**Comment by klecki** [comment]
Hmm, I can probably wrap it into intermediate template-function and call that one through macro thus removing the need to wrap the TYPE_SWITCH in macro.

**Comment by szalpal** [comment]
But you can put `TYPE_SWITCH` inside a `TYPE_SWITCH`, so it shouldn't be hard to put `TYPE_SWITCH` in something else...

File: dali/operators/expressions/expression_factory_instances/expression_impl_factory.h [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+  }
+
+
+
+/**
+ * @brief Factory function returning proper variant of implementation for `plus`
+ *        on CPUBackend for supplied input types and input kinds (Scalar/Tensor inputs),
+ *        specified in `expr`.
+ */
+std::unique_ptr<ExprImplBase> OpFactory(arithm_meta<ArithmeticOp::plus, CPUBackend>,
```
Comment (JanuszL) [inline dali/operators/expressions/expression_factory_instances/expression_impl_factory.h]: Maybe this signature should be generated by the macro too (so if you change it you don't need to change it everywhere)?

Comment (klecki) [reply dali/operators/expressions/expression_factory_instances/expression_impl_factory.h]: I wanted to add a docstring for this, and with a macro-generated function it's not possible.

**Pull request merged**